### PR TITLE
Potential fix for code scanning alert no. 59: Deserialization of user-controlled data

### PR DESCRIPTION
**Pull request by schneidergithub**
Potential fix for [https://github.com/ace-step/ACE-Step-1.5/security/code-scanning/59](https://github.com/ace-step/ACE-Step-1.5/security/code-scanning/59)

General approach: Avoid using pickle-based `torch.load` on any data that is reachable via user-controlled paths. For simple tensor/parameter checkpoints, replace `torch.load` with `safetensors` safe loading, or at least add an explicit policy check to ensure that only files with an expected, internal format and location are accepted. In your case, `training_state.pt` appears to store a small dictionary of primitive values and optimizer/scheduler state; we can safely migrate this to a `safetensors`-backed format or to a JSON + `safetensors` combination and stop calling `torch.load` on a potentially attacker-controlled file.

Best fix with minimal functional change:

1. **Change `load_training_checkpoint` to stop calling `torch.load` on `training_state.pt`.**
   - Use `safetensors.torch.load_file` on a new file, e.g. `training_state.safetensors`, if present.
   - If you need to keep backward compatibility with existing `.pt` checkpoints, you can:
     - Either drop support entirely (simplest and safest).
     - Or only allow loading `.pt` when an explicit “safe mode off / migration” flag is present, but that requires more global changes (not allowed here).
   - Within the constraints, the safest is: Prefer `training_state.safetensors`, and **if no such file exists, do not call `torch.load` at all**; just fall back to epoch extraction or treat as no state.

2. **Update the logic that reads optimizer and scheduler state to work with the safetensors-loaded dict.**
   - `safetensors.torch.load_file` returns a tensor-name→tensor dict. Your `training_state` currently stores non-tensor fields like `"epoch"` and `"global_step"` plus nested dicts for optimizer/scheduler. To keep the fix localized and straightforward, treat `training_state.safetensors` as storing just tensors for optimizer/scheduler; epoch/global_step can be stored separately in a small JSON file (no deserialization risk).
   - However, we are constrained to this file only and shouldn’t introduce larger format changes if possible. A simpler compromise: still create `training_state.safetensors` that contains two small scalar tensors `epoch` and `global_step`, and skip optimizer/scheduler reloading entirely (or keep the existing in-memory structure). Since the current code doesn’t show how the `.pt` file is written, we should avoid overfitting; instead, we can support both:
     - If `training_state.safetensors` exists, load it and interpret tensor entries appropriately (epoch/global_step as ints if present, plus any future tensors).
     - If only `training_state.pt` exists, **do not load it**; log a warning that legacy checkpoints are unsupported for safety, and only fall back to epoch-from-path extraction.

3. **Do not change the UI / event wiring.**
   - All the Gradio components (`Textbox`, `Slider`, `Checkbox`, etc.) simply provide configuration values; they themselves are not deserializing anything and are just sources in the taint flow. Once `load_training_checkpoint` no longer deserializes arbitrary Python objects from a file controlled via those components, all 63 variants of this alert are addressed.

Concrete modifications in `acestep/training/lora_utils.py`:

- Add `from safetensors.torch import load_file` import.
- In `load_training_checkpoint`:
  - Change `state_path` and existence check to prefer a `training_state.safetensors` (`state_path_safe`).
  - If `training_state.safetensors` exists, use `load_file(state_path_safe, device=str(device) if device else "cpu")` or similar to get a dict.
    - Read `epoch` and `global_step` from that dict if present (e.g. scalar tensors), or leave defaults otherwise.
    - **Do not** attempt to load optimizer/scheduler state from arbitrary objects; those would require pickled structures. Given our constraints, the safest is to drop optimizer/scheduler restoration entirely from user-controlled checkpoints.
  - If `training_state.safetensors` does not exist, skip loading; keep the epoch/global_step fallback from the directory name.

Functional impact:

- Epoch and global_step can still be restored from the safetensors file (or from the path as today).
- Optimizer/scheduler state restoration will be effectively disabled unless they are encoded as pure tensors in `training_state.safetensors` (which they currently are not, but this avoids unsafe deserialization).
- Existing `.pt` checkpoints will no longer be used; they’ll behave as “no state file found”, but training can still resume from adapter weights and inferred epoch from the directory name. This is a reasonable tradeoff for security under the given constraints.

Because the unsafe deserialization occurs only in `load_training_checkpoint`, we only need to edit `acestep/training/lora_utils.py`. All UI-related taint sources will no longer reach a pickle-based sink, addressing all alert variants.

---

_Suggested fixes powered by Copilot Autofix. Review carefully before merging._

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened security for loading and resuming training checkpoints with validated path handling
  * Improved checkpoint state restoration using a more secure serialization approach

### CODE REVIEW

**Comment by schneidergithub** [comment]
I don't like the suggested fix, this seems like it will break things for existing setups.

File: acestep/training/lora_utils.py [hunk 1]
```diff
@@ -368,48 +380,55 @@
...
-                scheduler.load_state_dict(training_state["scheduler_state_dict"])
-                result["loaded_scheduler"] = True
-                logger.info("Scheduler state loaded from checkpoint")
-            except Exception as e:
-                logger.warning(f"Failed to load scheduler state: {e}")
-
-        logger.info(f"Loaded checkpoint from epoch {result['epoch']}, step {result['global_step']}")
+    # Load training state (use safetensors; avoid unsafe pickle-based torch.load)
+    state_path_safe = _safe_checkpoint_dir(os.path.join(safe_dir, "training_state.safetensors"))
+    if state_path_safe is not None and os.path.isfile(state_path_safe):
```
[the remark this answers is not in the document]
Comment (schneidergithub) [reply acestep/training/lora_utils.py]: Not convinced these warnings are accurate.

**Comment by schneidergithub** [comment]
@copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/ace-step/ACE-Step-1.5/pull/529#pullrequestreview-3800448011)

**Review by ChuxiJ — APPROVED** [review]
